include/rdma/fi_eq.h

```diff
@@ -230,6 +231,21 @@ struct fi_cq_tagged_entry {
 	uint64_t		tag;
 };
 
+struct fi_cq_rpc_entry {
+	void			*op_context;
+	uint64_t		flags;
+	size_t			len;
+	void			*buf;
+	union {
+		uint64_t	data;
+		int		timeout;
+	};
```

Comment:
This change effectively blocks RPC from using CQ data.  That may be undesirable.  It forces RPC requests to match against a posted receive buffer, even if the request could be conveyed using only CQ data.

Since RPC requests arrive as untagged messages, they should have the same abilities as untagged messages, including carrying CQ data and supporting multi-recv.  It's reasonable to replace 'tag' with 'rpc_id', since RPC requests go through the untagged queues, but timeout should extend the structure.  Timeout may be useful as a completion field in some future change.

Reply:
I would expect an RPC request need more than 64 bits to describe the details of the request. It is reasonable to require an extra message body for the request. And, unlike msg/tagged send calls, we don't have a 'data' variant for rpc so why not go further to remove data from the `msg` variant as well.

Using the data field for the timeout value greatly simplifies the process of adding RPC support to existing providers.

Reply:
There's no data variant because it was excluded.  I'm questioning that decision.  Tagged and untagged messages (and even RMA) support CQ data.  Why exclude it from RPC?  We don't know how apps might use this API, an RPC request with a command but no input data could easily use just CQ data.  Or the input data might be transferred using some other mechanism, since RPC requires landing the request into a posted receive buffer.

The API is fixed nearly forever.  It's more important to get it right.